### PR TITLE
Config validation, waiting-until, and this repo's .mcp.json (TASK-036, 052, 056)

### PR DESCRIPTION
**Pull request by refined**
Three of the four tasks planned for this release. **TASK-053 (archiving `DONE.md`) is deferred** and stays in the backlog.

## TASK-036 — `config.json` is validated on load (P1)

Reproduced the failure before fixing it. With `states` given as plain strings, `migrateConfig` looks for `s.name === 'Rejected'` on a string, finds nothing, and appends an object to a string array; then `path.join(tasksDir, state.fileName)` throws `TypeError: The "path" argument must be of type string`.

`load()` no longer throws. Parsing is wrapped, and `normalizeStates` repairs entries naming a known state from `DEFAULT_STATES`. An entry that cannot be recognised costs the whole list rather than leaving a partial board, which would be worse than the default one.

Diagnostics are returned rather than logged, keeping core VS Code-free. The extension writes them to the existing **TaskPlanner** output channel with a *Show details* action — degrading silently to defaults is how someone ends up running settings they did not choose.

**This fixes more than the UI.** `load()` is called from `freshStore` in the MCP server, so until now a broken config failed *every tool call* with an opaque error. Verified end to end that the config which used to throw now answers `taskplanner_board` and reports three diagnostics on stderr.

## TASK-052 — tasks can say when they cannot start yet

The generated workflow tells agents to take the highest-priority task from NEXT, and a task had no way to say "not yet". In isotopy a P0 sits in NEXT blocked on an external quota until September; a compliant agent picks it and cannot do it.

A `waitingUntil` field rather than a `Blocked` state — a state would have reshaped every existing board and needed a `states` migration.

**Made it mean something rather than only storing it**, which is where `epic` went wrong before TASK-051: `structuredTask()` derives a `waiting` boolean for every tool so no client redoes the date comparison, the board marks such tasks, and the generated instructions tell agents to skip them whatever their priority.

`isWaiting` compares date-only strings, avoiding timezone drift from parsing into a `Date`. Waiting until *today* counts as available, and an unparseable value counts as **not** waiting — a typo must not hide work forever.

Caught while testing: the parser branch alone was not enough, because `flushTask` builds each `Task` field by field and silently dropped the value. The round-trip test is what surfaced it.

## TASK-056 — this repository finally has a `.mcp.json`

TASK-054 shipped the ability to write one and recorded that this repository did not get it, because `npx` could not resolve the package until it was published. It is published now.

Written by the shipped `writeMcpServerConfig()` rather than by hand, so the file is evidence the 2.2.0 feature works rather than a hand-made lookalike. Verified by launching exactly what it specifies: `npx -y @smekai/taskplanner` pulls 2.2.0 from the registry, reports 8 tools, and answers `taskplanner_board` against this repository's real board.

That also settles the question left open in TASK-054's plan — `npx` resolves the `taskplanner-mcp` bin despite its name differing from the package name, so the `-p` form is unnecessary.

**It has taken effect:** the `taskplanner_*` tools are now registered in this repository's sessions, so the board can be driven through them instead of by hand.

## Verification

- `npm run release:check` green: **162 tests**, up from 148. Version markers consistent, plugin validators and published-artifact smoke test pass.
- End-to-end runs against a live server for the `waitingUntil` round-trip (create with future and past dates, move, update, board marking, markdown) and for the broken-config path.
- Six malformed-config shapes exercised: bare-string states, a known state missing a field, an unrecognisable entry, truncated JSON, an empty file, a JSON array. None throw; `FileStore` works in every case.
- Agent instructions regenerated once at the end rather than per task, matching how 2.2.0 was done.

## On the version number

The branch is named `2.3.0`, but **nothing has been bumped to 2.3** — the pre-commit hook does patch bumps and the branch currently sits at **2.2.4**.

My reasoning for planning a minor was `waitingUntil`: a new markdown field, new tool inputs and new instruction text is an addition rather than a fix, and under semver that is a minor. TASK-036 alone would have been a patch, and TASK-056 does not affect the published package at all.

Now that TASK-053 is deferred, the release is smaller than planned, so this is worth a deliberate decision rather than inheriting my earlier one. Leaving it at 2.2.4 unless told otherwise; setting 2.3.0 is one command before merge.

## Deferred

**TASK-053** stays in the backlog. It is the one item in this release with a correctness hazard: `getMaxTaskIdNumber()` scans only configured states, so completed tasks moved into a file outside `states[]` would stop being seen and `nextId` could reissue their IDs. The groundwork for doing it safely is already in the codebase — `maxTaskIdNumber(raw, prefix)` and `FileStore.readRawContent()`, the same pair used for deferred states.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/core/config/configManager.ts [hunk 1]
```diff
@@ -31,9 +94,35 @@ export class ConfigManager {
...
+      if (!value || typeof value !== 'object' || Array.isArray(value)) {
+        report('config.json is not a JSON object; using defaults.');
+      } else {
+        parsed = value as Partial<TaskPlannerConfig>;
+      }
+    } catch (error) {
+      report(`config.json could not be parsed (${(error as Error).message}); using defaults.`);
+    }
+
+    const merged = { ...createDefaultConfig(), ...parsed };
```
Comment (refined) [inline src/core/config/configManager.ts]: [P1] Validate the whole config boundary here — this is also the clearest place to make the implementation shorter. The raw spread still trusts every field except `states`: for example, `{ "version": 3, "priorities": {} }` loads with zero diagnostics, then `createTask.ts` throws at `config.priorities.map(...)`. `tags`, `nextId`, `idPrefix`, `insertPosition`, and the booleans are likewise unchecked. One schema/decoder with defaults, passthrough for unknown keys, and a preprocess/transform for legacy state names could replace much of `isUsableState`/`normalizeStates`/the parse branching while guaranteeing the runtime `TaskPlannerConfig` type for every consumer.

File: src/core/config/configManager.ts [hunk 2]
```diff
@@ -31,9 +94,35 @@ export class ConfigManager {
     if (!fs.existsSync(this.configPath)) {
       return;
     }
-    const raw = fs.readFileSync(this.configPath, 'utf-8');
-    const parsed = JSON.parse(raw) as Partial<TaskPlannerConfig>;
-    this.config = { ...createDefaultConfig(), ...parsed };
+    this.diagnostics = [];
+    this.config = this.readFromDisk();
```
Comment (refined) [inline src/core/config/configManager.ts:184]: [P1] Preserve an unreadable config before returning writable defaults. `TaskStore.createTask()` calls this method, then `getNextId()` saves the resulting config; with `{ broken` on disk I reproduced the first create replacing it with a valid default v3 file. That turns the new fail-open read path into silent data loss before the user can use the reported diagnostic to repair the file. Track a failed-load/read-only state and reject mutations until repair, or save a backup/quarantine copy before any default config is written.

File: src/core/util/time.ts [hunk 3]
```diff
@@ -2,3 +2,22 @@
...
+/**
+ * True when a task is waiting on a date that has not arrived. Comparison is date-only and string
+ * based, which is safe for `YYYY-MM-DD` and avoids timezone drift from parsing into a Date.
+ * A task waiting until today counts as available.
+ *
+ * An unparseable value is treated as not waiting: a typo should not silently hide work forever.
+ */
+export function isWaiting(waitingUntil: string | undefined, today = currentDate()): boolean {
+  if (!waitingUntil) return false;
+  const date = waitingUntil.trim().slice(0, 10);
```
Comment (refined) [inline src/core/util/time.ts]: [P1] Validate a real calendar date before using lexical ordering. The regex accepts impossible values — `isWaiting('2026-99-99', '2026-08-27')` returns `true` — so exactly the typo this function promises not to hide can suppress a task forever; slicing also accepts arbitrary suffixes despite the `YYYY-MM-DD` contract. Also, `currentDate()` derives UTC rather than the user's local calendar day, making tasks available early west of UTC and late east of UTC around midnight. Parse and round-trip the year/month/day components, then compare against an explicitly local (or explicitly documented UTC) today.

**Review by refined — COMMENTED** [review]
Reviewed the config boundary and waiting-date behavior, then ran the complete release gate locally: all 162 tests, validators, build, and published-package smoke checks pass. The config decoder is the strongest opportunity to make this shorter and more complete; I also found two data-loss/date edge cases below.

**Comment by refined** [comment]
One additional cleanup request: please remove unnecessary source comments that only narrate or restate what the surrounding code already makes clear. Prefer clearer names or small extracted functions where needed, and keep comments only for non-obvious constraints or behavior that the code cannot communicate on its own.

File: src/core/config/configManager.ts (hunk 1 above)
Comment (refined) [reply src/core/config/configManager.ts]: Fixed in 7ea0a02. Reproduced first: `{ "version": 3, "priorities": {} }` loaded with zero diagnostics, then `config.priorities.map is not a function`. You were right that only `states` was being checked — the spread trusted everything else.

Replaced the raw spread with per-field decoders covering every declared field, so the object handed to consumers matches its type at runtime rather than only at compile time. Unknown keys still pass through untouched, so settings written by a newer TaskPlanner survive a round-trip through an older one — there is a test for that.

One place I went a different way than suggested: kept it hand-rolled rather than reaching for a schema library. `src/core/` imports nothing but `fs` and `path` today, and that property is load-bearing now that the same code ships as the `@smekai/taskplanner` library bundle. Pulling zod in for validation would inline it into that bundle for every consumer. The decoder table came out about the same length as what it replaced.

`normalizeStates` stayed separate rather than folding into the table: it repairs by *name* against `DEFAULT_STATES`, which is a different operation from the scalar checks, and inlining it made the table harder to read rather than shorter.

File: src/core/config/configManager.ts (hunk 2 above)
Comment (refined) [reply src/core/config/configManager.ts:184]: Fixed in 7ea0a02, and thank you for catching it — this was a regression my own TASK-036 fix introduced, which is worse than the crash it replaced. Reproduced exactly as you describe: a file containing `{ broken` was gone after a single `createTask`.

The raw bytes are now held from the failed read and written to `config.invalid-<timestamp>.json` before defaults overwrite anything, with a diagnostic naming the backup. If the backup itself cannot be written, the original is left in place rather than lost silently.

Took the quarantine option over read-only-until-repair deliberately. Refusing to save would leave `getNextId()` unable to persist `nextId`, so every allocation would return the same ID — trading silent data loss for silent board corruption, which seemed the worse of the two.

Two tests: a broken file leaves exactly one backup with the original bytes, and a readable file leaves none.

File: src/core/util/time.ts (hunk 3 above)
Comment (refined) [reply src/core/util/time.ts]: Fixed in 7ea0a02. Confirmed `isWaiting('2026-99-99', '2026-08-27')` returned `true` — the shape check passed and the task would have been suppressed indefinitely, which is precisely what the function documents it will not do. Now the components are parsed and round-tripped through a Date, so `2026-99-99`, `2026-02-31` and `2026-13-01` are all rejected.

`currentDate()` now derives the local calendar day rather than UTC.

One deliberate difference from your suggestion: a **trailing time** is still accepted. `**Updated:**` in the same files is written as `YYYY-MM-DD HH:mm`, so a user typing a date with extra precision is following the local convention rather than making a mistake, and rejecting it would silently make a blocked task available. What is now rejected is text glued directly to the date — `2026-09-03xyz` — rather than silently truncated. Tested both ways.

If you would rather have strict bare-date-only, say so and I will tighten it; the tradeoff is which direction the surprise goes.
